library/core/src/mem/mod.rs

```diff
@@ -726,7 +726,11 @@ pub const fn swap<T>(x: &mut T, y: &mut T) {
         // tends to copy the whole thing to stack rather than doing it one part
         // at a time, so instead treat them as one-element slices and piggy-back
         // the slice optimizations that will split up the swaps.
-        if size_of::<T>() / align_of::<T>() > 4 {
+        //
+        // Also, simple swaps of types bigger than 2 usizes cause unneccessary
+        // stack allocation: https://godbolt.org/z/6r1Pfznjf
+        // See also commit message.
+        if size_of::<T>() / align_of::<T>() > 4 || size_of::<T>() > 2 * size_of::<usize>() {
```

Comment:
If this check is going to be changed, I think it should actually be changed *more*, because with just this change it'll send a bunch of things to the `ptr` version which will end up in the loop that just calls `swap_simple` anyway, which would be better to just not bother (since it's much easier to optimize-out the constant `if` here than the one-iteration-after-inlining loop being called).

For example, the `size/align > 4` check didn't pick up things like `_m512`, so if this check expands, please add a new check like https://godbolt.org/z/PbMebWqod (probably in a separate file, to be able to pass the fancier target).

So if this helps then great, but I think it should pick up more parts of the check in `ptr` -- maybe don't bother for things with `align > usize`, for example.  Or experiment with whether `[usize; 4]` is actually fine (maybe try swapping a `Deque`?) and it's the non-power-of-two-ness of `Vec` that's really causing the problem, which might imply a different check here is better.

Reply:
There is already check for SIMD types: https://github.com/rust-lang/rust/blob/ada8c80bedb713b320af00aacab97d01d9cb5933/src/test/codegen/swap-simd-types.rs#L16
Do I need to add for `_m512` too?

Do I need separate test for VecDeque though? It does generate alloca like Vec without my changes and doesn't touch the stack with my changes. Seems to be exactly same logic as with Vec so Vec test covers it all.

Reply:
Ah, my mistake -- I was thinking the other test was __m128, and thus wouldn't be `> 2×usize`.

I still think that something like
```suggestion
        if size_of::<T>() / align_of::<T>() > 4 || (size_of::<T>() > 2 * size_of::<usize>() && align_of::<T>() <= align_of::<usize>()) {
```
would be good since the `ptr` version doesn't do anything special for highly-aligned `T` anyway.

(Or maybe parenthesized the other way?)

Reply:
Added check for alignment <= align of usize.